pr_data_wrangling.py

```diff
@@ -0,0 +1,154 @@
+import polars as pl
+import pandas as pd
+import numpy as np
+import os, glob
+import re
+import plotly.express as px
+import polars as pl
+
+# directories
+cwd = os.getcwd() # directory this file is saved in
+data_dir = os.path.join(cwd, "BCDC-Metadata") # BCDC data folder
+metadata_dir = os.path.join(data_dir, 'Sample-Inventory') # directory with specimen metadata
+project_dir = os.path.join(data_dir, 'Data-Collection-Inventory', '2022Q3_csvs')
+
+# using the data cleaned by Pam, which is stored separately in the current wd
+df = pd.read_csv(os.path.join(cwd, 'BCDC_Metadata_2022Q3_v4.csv'), encoding='unicode_escape')
+
+# clean up data and columns
+df.columns = df.columns.str.replace(r"\([^)]*\)","").str.strip() # get rid of (CV) in data columns
+df['Sample Type'] = df['Sample Type'].str.replace("[^A-Za-z0-9 ]+", " ").str.lower() # special characters into spaces, set all to lower - shouldn't have special characters in sample type
+df['Species'] = df['Species'].str.lower() # species could be converted to all lowercase as well
+df['Subspecimen Type'] = df['Subspecimen Type'].str.replace("[^A-Za-z0-9 ]+", " ").str.lower() # special characters into spaces, all lowercase
+df['Total Processed Subspecimens'] = df['Total Processed Subspecimens'].astype(str).str.replace(',', '').astype(float) # cell counts is numeric
+df['Sample ID'] = df['Sample ID'].astype(str) # so that IDs with all strings won't somehow get counted as numerics down the road
+df['Species'] = df['Species'].str.lower()
+df['Modality'] = df['Modality'].str.lower()
+# if either sample type of subspecimen type is NA, use the other column
+df['Sample Type'] = df['Sample Type'].fillna(df['Subspecimen Type'])
+df['Subspecimen Type'] = df['Subspecimen Type'].fillna(df['Sample Type'])
+# if no reported specimens, count as 0
+df['Total Processed Subspecimens'] = df['Total Processed Subspecimens'].fillna(0)
+### some columns are not getting considered at moment
+optional_cols = ['Age', 'Sex', 'Genotype', 'Species NCBI Taxonomy ID']
+df = df.drop(optional_cols, axis = 1)
+df = df.drop_duplicates()
+
+# fix the mc seq issue 
+df.loc[(df['Technique'].str.contains('mc-seq', case = False)) & (~df['Technique'].str.contains(';', case = False)), 'Technique'] = 'mC-seq2'
+
+#### Count for number of samples per categories
+## implemented 'species' as a category for now, more to be added
+sample_count_df = pd.DataFrame(df.groupby(['Data Collection', 'Metadata Submission', 'Species'], as_index = False).agg({'Sample ID': 'nunique'})).rename(columns = {'Data Collection':'data_collection_reference_id', 'Sample ID':'sample_count', 'Metadata Submission':'quarters', 'Sample Type': 'sample_name', 'Species':'species'})
```

Comment:
This line is 328 characters long, which can make it difficult to read for people with sight disabilities. The PEP 8  style guide recommends lines be 79 characters or less, but Black defaults to 10% more at 88 characters.

Some reading:
- PEP-8 line length: https://peps.python.org/pep-0008/#maximum-line-length
- Black line length: https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length

Here's an example of what Black will do to your code if you use it to auto-format your code:
```suggestion
sample_count_df = pd.DataFrame(
    df.groupby(
        ["Data Collection", "Metadata Submission", "Species"], as_index=False
    ).agg({"Sample ID": "nunique"})
).rename(
    columns={
        "Data Collection": "data_collection_reference_id",
        "Sample ID": "sample_count",
        "Metadata Submission": "quarters",
        "Sample Type": "sample_name",
        "Species": "species",
    }
)
```